webrtc/dtls/dtls_connection.nim

```diff
@@ -0,0 +1,222 @@
+
+proc dtlsHandshake*(
+    self: DtlsConn,
+    isServer: bool
+  ) {.async: (raises: [CancelledError, WebRtcError]).} =
+  var shouldRead = isServer
+  try:
+    while self.ctx.ssl.private_state != MBEDTLS_SSL_HANDSHAKE_OVER:
+      if shouldRead:
+        if isServer:
+          case self.raddr.family
+          of AddressFamily.IPv4:
+            mb_ssl_set_client_transport_id(self.ctx.ssl, self.raddr.address_v4)
+          of AddressFamily.IPv6:
+            mb_ssl_set_client_transport_id(self.ctx.ssl, self.raddr.address_v6)
+          else:
+            raise newException(WebRtcError, "DTLS - Remote address isn't an IP address")
+        let (data, _) = await self.conn.read()
+        self.dataRecv = data
+      self.sendFuture = nil
+      let res = mb_ssl_handshake_step(self.ctx.ssl)
+      if not self.sendFuture.isNil():
+        await self.sendFuture
+      shouldRead = false
+      if res == MBEDTLS_ERR_SSL_WANT_WRITE:
+        continue
+      elif res == MBEDTLS_ERR_SSL_WANT_READ:
+        shouldRead = true
+        continue
+      elif res == MBEDTLS_ERR_SSL_HELLO_VERIFY_REQUIRED:
+        mb_ssl_session_reset(self.ctx.ssl)
+        shouldRead = isServer
+        continue
+      elif res != 0:
+        raise newException(WebRtcError, "DTLS - " & $(res.mbedtls_high_level_strerr()))
+  except MbedTLSError as exc:
+    trace "Dtls handshake error", errorMsg = exc.msg
+    raise newException(WebRtcError, "DTLS - Handshake error", exc)
+
+proc close*(self: DtlsConn) {.async: (raises: [CancelledError, WebRtcError]).} =
+  ## Close a Dtls Connection
+  ##
+  if self.closed:
+    debug "Try to close an already closed DtlsConn"
+    return
+
+  self.closed = true
+  self.sendFuture = nil
+  # TODO: proc mbedtls_ssl_close_notify => template mb_ssl_close_notify in nim-mbedtls
+  let x = mbedtls_ssl_close_notify(addr self.ctx.ssl)
+  if not self.sendFuture.isNil():
+    await self.sendFuture
+  self.closeEvent.fire()
+
+proc write*(self: DtlsConn, msg: seq[byte]) {.async.} =
+  ## Write a message using mbedtls_ssl_write
+  ##
+  # Mbed-TLS will wrap the message properly and call `dtlsSend` callback.
+  # `dtlsSend` will write the message on the higher Stun connection.
+  if self.closed:
+    debug "Try to write on an already closed DtlsConn"
+    return
+  var buf = msg
+  try:
+    self.sendFuture = nil
+    let write = mb_ssl_write(self.ctx.ssl, buf)
+    if not self.sendFuture.isNil():
+      let sendFuture = self.sendFuture
+      await sendFuture
+    trace "Dtls write", msgLen = msg.len(), actuallyWrote = write
+  except MbedTLSError as exc:
+    trace "Dtls write error", errorMsg = exc.msg
+    raise exc
+
+proc read*(self: DtlsConn): Future[seq[byte]] {.async.} =
+  if self.closed:
+    debug "Try to read on an already closed DtlsConn"
+    return
+  var res = newSeq[byte](8192)
+  while true:
+    let (data, _) = await self.conn.read()
+    self.dataRecv = data
+    # TODO: Find a clear way to use the template `mb_ssl_read` without
```

Comment:
Could you please replace the TODOs with issues? In this way, we have better visibility.

Reply:
Those two TODOs are mbedtls related, as you suggested I opened issues there : https://github.com/status-im/nim-mbedtls/issues/2